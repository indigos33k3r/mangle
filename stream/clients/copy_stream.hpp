#ifndef __MANGLE_STREAM_COPYSTREAM_HPP_
#define __MANGLE_STREAM_COPYSTREAM_HPP_

#include "../stream.hpp"
#include <vector>

namespace Mangle {
namespace Stream {

  /** CopyStream::copy() is a convencience function that reads from
      one stream and write to another.

      It's wrapped as a header-only static struct function for easier
      compiler setup.

      You can either read the whole stream, or stop after a given
      number of bytes.
   */
  struct CopyStream
  {
    static size_t copy(StreamPtr input, StreamPtr output,
                       size_t stopAfter = 0)
    {
      assert(input->isReadable);
      assert(output->isWritable);

      size_t toRead = stopAfter;

      if(!toRead)
        {
          // Calculate input size
          if(input->hasSize && input->hasPosition)
            toRead = input->size() - input->tell();
        }

      // Tell the output that we are about to write. If toRead is
      // still zero at this point, it means 'read to end of stream'.
      if(toRead)
        output->reserve(toRead);

      // Can we use pointers?
      if(input->hasPtr && toRead)
        {
          // That makes things easy
          output->write(input->getPtr(toRead), toRead);
          return toRead;
        }

      // Read in increments
      std::vector<char> buf(0x10000);
      size_t written = 0;
      while(!input->eof())
        {
          size_t count = buf.size();
          if(toRead && count > toRead)
            count = toRead;
          count = input->read(&buf[0], count);
          written += output->write(&buf[0], count);

          if(toRead)
            {
              toRead -= count;
              if(toRead == 0)
                break;
            }                 
        }
      return written;
    }
  };
}}
#endif
