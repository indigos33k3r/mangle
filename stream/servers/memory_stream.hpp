#ifndef MANGLE_STREAM_MEMSERVER_H
#define MANGLE_STREAM_MEMSERVER_H

#include <assert.h>
#include "../stream.hpp"
#include <string.h>

namespace Mangle {
namespace Stream {

// Do this before the class declaration, since the class itself
// uses it.
class MemoryStream;
typedef boost::shared_ptr<MemoryStream> MemoryStreamPtr;

/** A Stream wrapping a memory buffer

    This will create a fully seekable stream out of any pointer/length
    pair you give it.
 */
class MemoryStream : public Stream
{
  const void *data;
  size_t length, pos;

 public:
 MemoryStream(const void *ptr, size_t len)
   : data(ptr), length(len), pos(0)
  {
    isReadable = true;
    isSeekable = true;
    hasPosition = true;
    hasSize = true;
    hasPtr = true;
  }

  static MemoryStreamPtr Open(const void *ptr, size_t len)
  { return MemoryStreamPtr(new MemoryStream(ptr, len)); }

  MemoryStream()
    : data(NULL), length(0), pos(0)
  {
    isReadable = true;
    isSeekable = true;
    hasPosition = true;
    hasSize = true;
    hasPtr = true;
  }

  size_t read(void *buf, size_t count)
  {
    assert(data != NULL);
    assert(pos <= length);

    // Don't read more than we have
    if(count > (length - pos))
      count = length - pos;

    // Copy data
    if(count)
      memcpy(buf, ((char*)data)+pos, count);

    // aaand remember to increase the count
    pos += count;

    return count;
  }

  void seek(size_t _pos)
  {
    pos = _pos;
    if(pos > length)
      pos = length;
  }

  size_t tell() const { return pos; }
  size_t size() const { return length; }
  bool eof() const { return pos == length; }

  const void *getPtr() { return data; }
  const void *getPtr(size_t size)
    {
      // This variant of getPtr must move the position pointer
      size_t opos = pos;
      pos += size;
      if(pos > length) pos = length;
      return ((char*)data)+opos;
    }
  const void *getPtr(size_t pos, size_t size)
    {
      if(pos > length) pos = length;
      return ((char*)data)+pos;
    }

  // New members in MemoryStream:

  /// Set a new buffer and length. This will rewind the position to zero.
  void set(const void* _data, size_t _length)
  {
    data = _data;
    length = _length;
    pos = 0;
  }

  /// Clone this memory stream
  /** Make a new stream of the same buffer. If setPos is true, we also
      set the clone's position to be the same as ours.

      No memory is copied during this operation, the new stream is
      just another 'view' into the same shared memory buffer.
  */
  MemoryStreamPtr clone(bool setPos=false) const
  {
    MemoryStreamPtr res(new MemoryStream(data, length));
    if(setPos) res->seek(pos);
    return res;
  }
};

}} // namespaces
#endif
