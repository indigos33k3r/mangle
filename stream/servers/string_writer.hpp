#ifndef MANGLE_STREAM_STRINGWRITER_H
#define MANGLE_STREAM_STRINGWRITER_H

#include "../stream.hpp"
#include <string>

namespace Mangle {
namespace Stream {

/** A stream that writes to an std::string. You provide the string
    yourself, so if you are writing large amounts of data you can
    optimize by calling string::reserve() first. You can also call
    string::reserve() indirectly through Stream::reserve().

    The data is appended to any existing data in the string.
 */
class StringWriter : public Stream
{
  std::string &out;

 public:
  StringWriter(std::string &_out)
    : out(_out)
  {
    isSeekable = false;
    hasPosition = false;
    hasSize = false;
    isWritable = true;
    isReadable = false;
  }

  size_t write(const void* buf, size_t len)
  {
    out.append((const char*)buf, len);
    return len;
  }

  void reserve(size_t count)
  {
    out.reserve(out.size() + count);
  }
};

typedef boost::shared_ptr<StringWriter> StringWriterPtr;

}} // namespaces
#endif
