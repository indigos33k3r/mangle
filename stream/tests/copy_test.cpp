#include <iostream>
#include <string.h>

#include "../servers/memory_stream.hpp"
#include "../servers/string_writer.hpp"
#include "../clients/copy_stream.hpp"

using namespace Mangle::Stream;
using namespace std;

void test(Stream *inp, size_t lim=0)
{
  string out;

  cout << "lim=" << lim << " size=" << inp->size()
       << " pos=" << inp->tell() << " left="
       << (inp->size() - inp->tell());
  lim = CopyStream::copy(StreamPtr(inp), StringWriter::Open(out), lim);
  cout << " wrote=" << lim << " result='" << out << "'\n";
}

int main()
{
  Stream *s = new MemoryStream("hello world", 11);
  test(s);
  s = new MemoryStream("hello world", 11);
  s->hasPtr = false;
  test(s);
  s = new MemoryStream("hello world", 11);
  s->hasSize = false;
  s->hasPtr = false;
  test(s);
  s = new MemoryStream("hello world", 11);
  s->hasSize = false;
  s->hasPtr = false;
  s->seek(4);
  test(s);
  s = new MemoryStream("hello world", 11);
  s->seek(4);
  test(s);
  s = new MemoryStream("hello world", 11);
  s->hasSize = false;
  s->hasPtr = false;
  s->seek(1);
  test(s,6);
  s = new MemoryStream("hello world", 11);
  s->hasPtr = false;
  s->seek(1);
  test(s,6);
  s = new MemoryStream("hello world", 11);
  s->seek(1);
  test(s,6);

  return 0;
}
