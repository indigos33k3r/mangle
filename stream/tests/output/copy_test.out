lim=0 size=11 pos=0 left=11 wrote=11 result='hello world'
lim=0 size=11 pos=0 left=11 wrote=11 result='hello world'
lim=0 size=11 pos=0 left=11 wrote=11 result='hello world'
lim=0 size=11 pos=4 left=7 wrote=7 result='o world'
lim=0 size=11 pos=4 left=7 wrote=7 result='o world'
lim=6 size=11 pos=1 left=10 wrote=6 result='ello w'
lim=6 size=11 pos=1 left=10 wrote=6 result='ello w'
lim=6 size=11 pos=1 left=10 wrote=6 result='ello w'
